internal/provider/provider.go

```diff
@@ -0,0 +1,309 @@
+package provider
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	runclient "github.com/fluxcd/pkg/runtime/client"
+	"github.com/fluxcd/pkg/ssa"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/provider"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/mitchellh/go-homedir"
+	"k8s.io/apimachinery/pkg/api/meta"
+	apimachineryschema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/rest"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/fluxcd/terraform-provider-flux/internal/utils"
+)
+
+type providerData struct {
+	Host                  types.String `tfsdk:"host"`
+	Username              types.String `tfsdk:"username"`
+	Password              types.String `tfsdk:"password"`
+	Insecure              types.Bool   `tfsdk:"insecure"`
+	ClientCertificate     types.String `tfsdk:"client_certificate"`
+	ClientKey             types.String `tfsdk:"client_key"`
+	ClusterCACertificate  types.String `tfsdk:"cluster_ca_certificate"`
+	ConfigPaths           types.Set    `tfsdk:"config_paths"`
+	ConfigPath            types.String `tfsdk:"config_path"`
+	ConfigContext         types.String `tfsdk:"config_context"`
+	ConfigContextAuthInfo types.String `tfsdk:"config_context_auth_info"`
+	ConfigContextCluster  types.String `tfsdk:"config_context_cluster"`
+	Token                 types.String `tfsdk:"token"`
+	ProxyURL              types.String `tfsdk:"proxy_url"`
+}
+
+var _ provider.Provider = &fluxProvider{}
+var _ provider.ProviderWithMetadata = &fluxProvider{}
+
+type fluxProvider struct {
+	version     string
+	fluxVersion string
+	manager     *ssa.ResourceManager
+	rcg         *RESTClientGetter
+	kubeClient  client.WithWatch
+}
+
+func New(version, fluxVersion string) func() provider.Provider {
+	return func() provider.Provider {
+		return &fluxProvider{
+			version:     version,
+			fluxVersion: fluxVersion,
+		}
+	}
+}
+
+func (p *fluxProvider) Metadata(ctx context.Context, req provider.MetadataRequest, resp *provider.MetadataResponse) {
+	resp.TypeName = "flux"
+	resp.Version = p.version
+}
+
+func (p *fluxProvider) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diagnostics) {
+	return tfsdk.Schema{
```

Comment:
Do you think it might make sense to nest these under a `kubernetes` subschema resource so the configuration would look something like this:
```
provider "flux" {
  kubernetes {
    config_path = "~/.kube/config"
  }
}
```

See here for "inspiration": https://github.com/hashicorp/terraform-provider-helm/blob/main/helm/provider.go#L149

Reply:
I think this is a good idea. Even if there are no other provider parameters right now, having them in their own objects makes it easier for the future.

Reply:
Sadly this is not possible as mixing blocks and nested attributes is not supported between the new and old provider frameworks. I think we should go ahead with the way the schema looks today and then at a later time when the old datasources are moved, attempt to move the configuration.

Reply:
I see, thanks for checking. It does make sense to wait until then in this case.